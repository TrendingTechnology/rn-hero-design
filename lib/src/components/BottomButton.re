open ReactNative;

[@bs.get] external getColorProperty: Style.t => Color.t = "color";

module Wrapper = {
  [@react.component]
  let make = (~enabled, ~onPress=?, ~style, ~children) => {
    switch (enabled, onPress) {
    | (true, Some(onPress)) =>
      <TouchableOpacity onPress style> children </TouchableOpacity>
    | (_, _) => <View style> children </View>
    };
  };
};

let emptyStyle = Style.style();

[@genType]
[@react.component]
let make =
    (
      ~text: string="",
      ~onPress,
      ~disabled=false,
      ~loading=false,
      ~forceInset=?,
      ~wrapperStyle=emptyStyle,
      ~buttonStyle=emptyStyle,
      ~textStyle=emptyStyle,
      ~theme=Hero_Theme.default,
    ) => {
  let (showKeyboard, setShowKeyboard) = React.useState(() => false);
  let hasHomeBar = Helpers.hasHomeBar();
  let themeWrapperStyle =
    switch (forceInset, showKeyboard, hasHomeBar) {
    | (Some("always"), _, _) => theme##bottomButton##wrapperWithHomeBar
    | (Some("never"), _, _) => theme##bottomButton##wrapperWithoutHomeBar
    | (_, false, true) => theme##bottomButton##wrapperWithHomeBar
    | (_, _, _) => theme##bottomButton##wrapperWithoutHomeBar
    };

  React.useEffect0(() => {
    let listener =
      Keyboard.addListener(`keyboardWillShow, _ => setShowKeyboard(_ => true));
    Some(() => EventSubscription.remove(listener));
  });

  React.useEffect0(() => {
    let listener =
      Keyboard.addListener(`keyboardWillHide, _ =>
        setShowKeyboard(_ => false)
      );
    Some(() => EventSubscription.remove(listener));
  });

  <Wrapper
    enabled={!loading && !disabled}
    style={StyleSheet.flatten([|
      theme##bottomButton##wrapper,
      themeWrapperStyle,
      disabled ? theme##bottomButton##disabledWrapper : emptyStyle,
      wrapperStyle,
    |])}
    ?onPress>
    <View
      style={StyleSheet.flatten([|theme##bottomButton##button, buttonStyle|])}>
      {switch (loading, disabled) {
       | (true, false) =>
         <ActivityIndicator
           size=ActivityIndicator.Size.small
           color={getColorProperty(theme##bottomButton##loadingIndicator)}
         />
       | (_, _) =>
         <Text
           style={StyleSheet.flatten([|
             theme##bottomButton##text,
             disabled ? theme##bottomButton##disabledText : emptyStyle,
             textStyle,
           |])}>
           text->React.string
         </Text>
       }}
    </View>
  </Wrapper>;
};

[@genType]
let default = Helpers.injectTheme(make);
