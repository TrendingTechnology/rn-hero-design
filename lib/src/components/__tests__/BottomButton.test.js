import React from 'react';
import BottomButton from '../BottomButton.gen';
import { fireEvent, render } from '@testing-library/react-native';

describe('BottomButton', () => {
  it.each`
    element
    ${(<BottomButton text="Button" onPress={() => {}} />)}
    ${(<BottomButton text="Button" onPress={() => {}} disabled />)}
    ${(<BottomButton text="Button" onPress={() => {}} loading />)}
    ${(<BottomButton text="Button" onPress={() => {}} forceInset="always" />)}
    ${(<BottomButton text="Button" onPress={() => {}} forceInset="never" />)}
  `('renders correctly', async ({ element }) => {
    const { baseElement } = render(element);
    expect(baseElement).toMatchSnapshot();
  });

  it('should feedback when users press', async () => {
    const callback = jest.fn();
    const { getByText } = render(
      <BottomButton text="Tap here" onPress={callback} />,
    );

    const button = getByText('Tap here');
    fireEvent.press(button);
    expect(callback).toBeCalledTimes(1);
  });

  it('should not feedback when disabled', async () => {
    const callback = jest.fn();
    const { getByText } = render(
      <BottomButton disabled text="Tap here" onPress={callback} />,
    );

    const button = getByText('Tap here');
    fireEvent.press(button);
    expect(callback).toBeCalledTimes(0);
  });
});
