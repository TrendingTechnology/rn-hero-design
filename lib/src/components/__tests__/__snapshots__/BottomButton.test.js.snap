// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BottomButton renders correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      style={
        Object {
          "backgroundColor": "#1dbeee",
          "borderRadius": 24.375,
          "marginBottom": 16.25,
          "marginHorizontal": 16.25,
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 48.75,
            "justifyContent": "center",
            "paddingHorizontal": 16.25,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fcfcfc",
              "fontFamily": "Proxima Nova",
              "fontSize": 16.25,
              "lineHeight": 16.25,
            }
          }
          value="Button"
        >
          Button
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`BottomButton renders correctly 2`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#d9dbdf",
          "borderRadius": 24.375,
          "marginBottom": 16.25,
          "marginHorizontal": 16.25,
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 48.75,
            "justifyContent": "center",
            "paddingHorizontal": 16.25,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fcfcfc",
              "fontFamily": "Proxima Nova",
              "fontSize": 16.25,
              "lineHeight": 16.25,
            }
          }
          value="Button"
        >
          Button
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`BottomButton renders correctly 3`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#1dbeee",
          "borderRadius": 24.375,
          "marginBottom": 16.25,
          "marginHorizontal": 16.25,
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 48.75,
            "justifyContent": "center",
            "paddingHorizontal": 16.25,
          }
        }
      >
        <ActivityIndicator
          animating={true}
          color="#fcfcfc"
          hidesWhenStopped={true}
          size="small"
        />
      </View>
    </View>
  </View>
</View>
`;

exports[`BottomButton renders correctly 4`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      style={
        Object {
          "backgroundColor": "#1dbeee",
          "borderRadius": 24.375,
          "marginBottom": 0,
          "marginHorizontal": 16.25,
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 48.75,
            "justifyContent": "center",
            "paddingHorizontal": 16.25,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fcfcfc",
              "fontFamily": "Proxima Nova",
              "fontSize": 16.25,
              "lineHeight": 16.25,
            }
          }
          value="Button"
        >
          Button
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;

exports[`BottomButton renders correctly 5`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <TouchableOpacity
      style={
        Object {
          "backgroundColor": "#1dbeee",
          "borderRadius": 24.375,
          "marginBottom": 16.25,
          "marginHorizontal": 16.25,
          "paddingBottom": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "height": 48.75,
            "justifyContent": "center",
            "paddingHorizontal": 16.25,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#fcfcfc",
              "fontFamily": "Proxima Nova",
              "fontSize": 16.25,
              "lineHeight": 16.25,
            }
          }
          value="Button"
        >
          Button
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;
